core/src/main/java/org/apache/iceberg/BaseFilesTable.java

```diff
@@ -0,0 +1,99 @@
+import org.apache.iceberg.expressions.Expression;
+import org.apache.iceberg.expressions.ManifestEvaluator;
+import org.apache.iceberg.expressions.Projections;
+import org.apache.iceberg.io.CloseableIterable;
+import org.apache.iceberg.relocated.com.google.common.collect.Sets;
+import org.apache.iceberg.types.TypeUtil;
+import org.apache.iceberg.types.Types.StructType;
+
+/**
+ * Base class logic for files metadata tables
+ */
+abstract class BaseFilesTable extends BaseMetadataTable {
+
+  BaseFilesTable(TableOperations ops, Table table, String name) {
+    super(ops, table, name);
+  }
+
+  @Override
+  public Schema schema() {
+    StructType partitionType = Partitioning.partitionType(table());
+    Schema schema = new Schema(DataFile.getType(partitionType).fields());
+    if (partitionType.fields().size() < 1) {
+      // avoid returning an empty struct, which is not always supported. instead, drop the partition field
+      return TypeUtil.selectNot(schema, Sets.newHashSet(DataFile.PARTITION_ID));
+    } else {
+      return schema;
+    }
+  }
+
+  abstract static class BaseFilesTableScan extends BaseMetadataTableScan {
+    private final Schema fileSchema;
+    private final MetadataTableType type;
+
+    protected BaseFilesTableScan(TableOperations ops, Table table, Schema fileSchema, MetadataTableType type) {
+      super(ops, table, fileSchema);
+      this.fileSchema = fileSchema;
+      this.type = type;
+    }
+
+    protected BaseFilesTableScan(TableOperations ops, Table table, Schema schema, Schema fileSchema,
+                           TableScanContext context, MetadataTableType type) {
+      super(ops, table, schema, context);
+      this.fileSchema = fileSchema;
+      this.type = type;
+    }
+
+    @Override
+    public TableScan appendsBetween(long fromSnapshotId, long toSnapshotId) {
+      throw new UnsupportedOperationException(
+          String.format("Cannot incrementally scan table of type %s", type.name()));
+    }
+
+    @Override
+    public TableScan appendsAfter(long fromSnapshotId) {
+      throw new UnsupportedOperationException(
+          String.format("Cannot incrementally scan table of type %s", type.name()));
+    }
+
+    protected CloseableIterable<ManifestFile> filterManifests(List<ManifestFile> manifests,
+                                                              Expression rowFilter,
+                                                              boolean caseSensitive) {
+      CloseableIterable<ManifestFile> manifestIterable = CloseableIterable.withNoopClose(manifests);
+
+      // use an inclusive projection to remove the partition name prefix and filter out any non-partition expressions
+      Expression partitionFilter = Projections
+          .inclusive(
+              transformSpec(fileSchema, table().spec(), PARTITION_FIELD_PREFIX),
+              caseSensitive)
+          .project(rowFilter);
+
+      ManifestEvaluator manifestEval = ManifestEvaluator.forPartitionFilter(
```

Comment:
It is not related to this PR but this logic may not be correct. We should create an issue and follow up. We always pass the current spec. Instead, we should use the spec to which our particular manifest belongs to.

Reply:
In addition, we should support predicate pushdown in ALL_DATA_FILES.

Reply:
> It is not related to this PR but this logic may not be correct. We should create an issue and follow up. We always pass the current spec. Instead, we should use the spec to which our particular manifest belongs to.

Made the issue:  https://github.com/apache/iceberg/issues/4271

Reply:
> In addition, we should support predicate pushdown in ALL_DATA_FILES.

Yea this is on my list, I was thinking after this base class it will make this thing much easier.